### ISSUE TITLE
OpenSSL 3.2.0-alpha1 04-test_bio_dgram fails on NonStop

### ISSUE THREAD

**Issue by rsbeckerca** [openssl/openssl#22013]
A significant number of tests are failing on NonStop. I have not included the detailed test log because of the size. Configuration is at the bottom. I can split this if they have different root causes:

Portion of the log:
```
04-test_bio_dgram.t .....................
        # INFO:  @ /home/randall/openssl/test/bio_dgram_test.c:125
        # # Testing with AF_INET6, local=0
        #
        # ERROR: (int) 'fd1 >= 0' failed @ /home/randall/openssl/test/bio_dgram_test.c:171
        # [-1] compared to [0]
        # 00000000:error:80001011:system library:BIO_socket:Address family not supported:/home/randall/openssl/crypto/bio/bio_sock2.c:51:calling socket()
        # 00000000:error:10000076:BIO routines:BIO_socket:unable to create socket:/home/randall/openssl/crypto/bio/bio_sock2.c:53:
        # OPENSSL_TEST_RAND_ORDER=1694107046
        not ok 2 - iteration 2
# ------------------------------------------------------------------------------
        # INFO:  @ /home/randall/openssl/test/bio_dgram_test.c:125
        # # Testing with AF_INET6, local=1
        #
        # ERROR: (int) 'fd1 >= 0' failed @ /home/randall/openssl/test/bio_dgram_test.c:171
        # [-1] compared to [0]
        # 00000000:error:80001011:system library:BIO_socket:Address family not supported:/home/randall/openssl/crypto/bio/bio_sock2.c:51:calling socket()
        # 00000000:error:10000076:BIO routines:BIO_socket:unable to create socket:/home/randall/openssl/crypto/bio/bio_sock2.c:53:
        # OPENSSL_TEST_RAND_ORDER=1694107046
        not ok 4 - iteration 4
# ------------------------------------------------------------------------------
    # OPENSSL_TEST_RAND_ORDER=1694107046
    not ok 1 - test_bio_dgram
# ------------------------------------------------------------------------------
        # ERROR: (int) 'i >= 9' failed @ /home/randall/openssl/test/bio_dgram_test.c:565
        # [6] compared to [9]
        # 00000000:error:10000070:BIO routines:dgram_pair_write:non-fatal or transient error:/home/randall/openssl/crypto/bio/bss_dgram_pair.c:1271:
        # OPENSSL_TEST_RAND_ORDER=1694107046
        not ok 5 - iteration 1
# ------------------------------------------------------------------------------
        # ERROR: (int) 'i >= 9' failed @ /home/randall/openssl/test/bio_dgram_test.c:565
        # [6] compared to [9]
        # 00000000:error:10000070:BIO routines:dgram_pair_write:non-fatal or transient error:/home/randall/openssl/crypto/bio/bss_dgram_pair.c:1271:
        # OPENSSL_TEST_RAND_ORDER=1694107046
        not ok 6 - iteration 2
# ------------------------------------------------------------------------------
    # OPENSSL_TEST_RAND_ORDER=1694107046
    not ok 2 - test_bio_dgram_pair
# ------------------------------------------------------------------------------
../../util/wrap.pl ../../test/bio_dgram_test => 255
not ok 1 - running bio_dgram_test
04-test_bio_dgram.t ..................... 1/? ----------------------------------
#   Failed test 'running bio_dgram_test'
#   at /home/randall/openssl/util/perl/OpenSSL/Test/Simple.pm line 77.
04-test_bio_dgram.t ..................... Dubious, test returned 1 (wstat 256, 0
```
**Configuration:**
```
Command line (with current working directory = .):

    perl ./Configure nonstop-nsx_64 --with-rand-seed=rdcpu enable-fips

Perl information:

    perl
    5.30.3 for NSV-D-nonstop_kernel

Enabled features:

    acvp-tests
    apps
    argon2
    aria
    async
    autoalginit
    autoerrinit
    autoload-config
    bf
    blake2
    bulk
    cached-fetch
    camellia
    capieng
    cast
    chacha
    cmac
    cmp
    cms
    comp
    ct
    deprecated
    des
    dgram
    dh
    docs
    dsa
    dso
    dtls
    dynamic-engine
    ec
    ec2m
    ecdh
    ecdsa
    ecx
    egd
    engine
    err
    filenames
    fips
    fips-securitychecks
    gost
    http
    idea
    legacy
    loadereng
    md4
    mdc2
    module
    multiblock
    nextprotoneg
    ocb
    ocsp
    padlockeng
    pic
    pinshared
    poly1305
    posix-io
    psk
    quic
    rc2
    rc4
    rdrand
    rfc3779
    rmd160
    scrypt
    secure-memory
    seed
    shared
    siphash
    siv
    sm2
    sm2-precomp
    sm3
    sm4
    sock
    srp
    srtp
    sse2
    ssl
    ssl-trace
    static-engine
    stdio
    tests
    tls
    ts
    ui-console
    whirlpool
    tls1
    tls1-method
    tls1_1
    tls1_1-method
    tls1_2
    tls1_2-method
    tls1_3
    dtls1
    dtls1-method
    dtls1_2
    dtls1_2-method

Disabled features:

    afalgeng            [not-linux]      OPENSSL_NO_AFALGENG
    asan                [default]        OPENSSL_NO_ASAN
    asm                 [no asm_arch]    OPENSSL_NO_ASM
    brotli              [default]        OPENSSL_NO_BROTLI
    brotli-dynamic      [default]        OPENSSL_NO_BROTLI_DYNAMIC
    buildtest-c++       [default]
    winstore            [not-windows]    OPENSSL_NO_WINSTORE
    crypto-mdebug       [default]        OPENSSL_NO_CRYPTO_MDEBUG
    default-thread-pool [cascade]        OPENSSL_NO_DEFAULT_THREAD_POOL
    devcryptoeng        [default]        OPENSSL_NO_DEVCRYPTOENG
    ec_nistp_64_gcc_128 [default]        OPENSSL_NO_EC_NISTP_64_GCC_128
    external-tests      [default]        OPENSSL_NO_EXTERNAL_TESTS
    fuzz-afl            [default]        OPENSSL_NO_FUZZ_AFL
    fuzz-libfuzzer      [default]        OPENSSL_NO_FUZZ_LIBFUZZER
    ktls                [default]        OPENSSL_NO_KTLS
    makedepend          [unavailable]
    md2                 [default]        OPENSSL_NO_MD2 (skip crypto/md2)
    msan                [default]        OPENSSL_NO_MSAN
    rc5                 [default]        OPENSSL_NO_RC5 (skip crypto/rc5)
    sctp                [default]        OPENSSL_NO_SCTP
    tfo                 [default]        OPENSSL_NO_TFO
    thread-pool         [cascade]        OPENSSL_NO_THREAD_POOL
    threads             [config]
    trace               [default]        OPENSSL_NO_TRACE
    ubsan               [default]        OPENSSL_NO_UBSAN
    unit-test           [default]        OPENSSL_NO_UNIT_TEST
    uplink              [no uplink_arch] OPENSSL_NO_UPLINK
    weak-ssl-ciphers    [default]        OPENSSL_NO_WEAK_SSL_CIPHERS
    zlib                [default]        OPENSSL_NO_ZLIB
    zlib-dynamic        [default]        OPENSSL_NO_ZLIB_DYNAMIC
    zstd                [default]        OPENSSL_NO_ZSTD
    zstd-dynamic        [default]        OPENSSL_NO_ZSTD_DYNAMIC
    ssl3                [default]        OPENSSL_NO_SSL3
    ssl3-method         [default]        OPENSSL_NO_SSL3_METHOD

Config target attributes:

    AR => "ar",
    ARFLAGS => "qc",
    CC => "c99",
    HASHBANGPERL => "/usr/bin/env perl",
    RANLIB => "ranlib",
    RC => "windres",
    bn_ops => "SIXTY_FOUR_BIT",
    build_file => "Makefile",
    build_scheme => [ "unified", "unix" ],
    cflags => "-g -O2 -Wextensions -Wnowarn=203,220,272,734,770,1506 -Wbuild_neutral_library -Wverbose -Wtarget=tns/x -Wsystype=oss -Wlp64 -WIEEE_float",
    cppflags => "",
    defines => [ "OPENSSL_BUILDING_OPENSSL", "OPENSSL_VPROC=\$(OPENSSL_VPROC)", "_XOPEN_SOURCE", "_XOPEN_SOURCE_EXTENDED=1", "_TANDEM_SOURCE", "__NSK_OPTIONAL_TYPES__", "B_ENDIAN" ],
    disable => [ "threads" ],
    dso_scheme => "DLFCN",
    enable => [ "egd" ],
    ex_libs => "-lrld",
    includes => [  ],
    lflags => "-Wxld=\"-set systype oss\" -Wxld=\"-set data_model lp64\" -Wxld=\"-set floattype ieee_float\"",
    lib_cflags => "",
    lib_cppflags => "",
    lib_defines => [  ],
    module_cflags => "",
    module_cxxflags => undef,
    module_ldflags => "-Wshared",
    multibin => "64",
    multilib => "64",
    perl => "/usr/bin/perl",
    perl_platform => "Unix",
    shared_argfileflag => "-Wxld_obey=",
    shared_cflag => "",
    shared_defflag => "-Wxld_obey=",
    shared_defines => [  ],
    shared_extension => ".so",
    shared_ldflag => "-Wshared",
    shared_rcflag => "",
    shared_target => "nonstop-shared",
    sys_id => "TANDEM",
    thread_defines => [  ],
    thread_scheme => "(unknown)",
    unistd => "<unistd.h>",

Recorded environment:

    AR =
    ARFLAGS =
    AS =
    ASFLAGS =
    BUILDFILE =
    CC =
    CFLAGS =
    CPP =
    CPPDEFINES =
    CPPFLAGS =
    CPPINCLUDES =
    CROSS_COMPILE =
    CXX =
    CXXFLAGS =
    HASHBANGPERL =
    LD =
    LDFLAGS =
    LDLIBS =
    MT =
    MTFLAGS =
    OPENSSL_LOCAL_CONFIG_DIR =
    PERL =
    RANLIB =
    RC =
    RCFLAGS =
    RM =
    WINDRES =
    __CNF_CFLAGS =
    __CNF_CPPDEFINES =
    __CNF_CPPFLAGS =
    __CNF_CPPINCLUDES =
    __CNF_CXXFLAGS =
    __CNF_LDFLAGS =
    __CNF_LDLIBS =

Makevars:

    AR              = ar
    ARFLAGS         = qc
    ASFLAGS         =
    CC              = c99
    CFLAGS          =
    CPPDEFINES      =
    CPPFLAGS        =
    CPPINCLUDES     =
    CXXFLAGS        =
    HASHBANGPERL    = /usr/bin/env perl
    LDFLAGS         =
    LDLIBS          =
    PERL            = perl
    RANLIB          = ranlib
    RC              = windres
    RCFLAGS         =

NOTE: These variables only represent the configuration view.  The build file
template may have processed these variables further, please have a look at the
build file for more exact data:
    Makefile

build file:

    Makefile

build file templates:

    Configurations/common0.tmpl
    Configurations/unix-Makefile.tmpl
```

**Comment by rsbeckerca** [comment]
The `80-test_cmp_http.t` test seems to have a problem:
```
# The results of this test will end up in test-runs/test_cmp_http
Starting1..6
Current directory is /home/randall/openssl/test/recipes/80-test_cmp_http_data/Mock
Launching Mock server: ../../../../util/wrap.pl ../../../../apps/openssl cmp -config server.cnf
Mock server PID=1409286238
cmp: error setting accept on port 0
00000000:error:80001011:system library:BIO_socket:Address family not supported:/home/randall/openssl/crypto/bio/bio_sock2.c:51:calling socket()
00000000:error:10000076:BIO routines:BIO_socket:unable to create socket:/home/randall/openssl/crypto/bio/bio_sock2.c:53:
00000000:error:80001011:system library:acpt_state:Address family not supported:/home/randall/openssl/crypto/bio/bss_acpt.c:233:calling socket(::, 0)
00000000:error:10000076:BIO routines:acpt_state:unable to create socket:/home/randall/openssl/crypto/bio/bss_acpt.c:236:
Mock server output: cmp_main:/home/randall/openssl/apps/cmp.c:3080:CMP info: using section(s) 'cmp' of OpenSSL configuration file 'server.cnf'
Killing Mock server with PID=0Killed
```
The server cannot have a PID=0 (that is the process itself), so appears to be causing the perl test wrapper process to terminate.

**Comment by rsbeckerca** [comment]
Moved `80-test_cmp_http` to #22014

**Comment by rsbeckerca** [comment]
There are some considerations on the platform that may apply here. Binding to specific sockets from 0-1024 are restricted to root only. I cannot change that. This applies to both UDP and TCP. I am wondering whether that is the cause of both errors above.

**Comment by hlandau** [comment]
> There are some considerations on the platform that may apply here. Binding to specific sockets from 0-1024 are restricted to root only. I cannot change that. This applies to both UDP and TCP. I am wondering whether that is the cause of both errors above.

It seems unlikely since the same restrictions apply to Linux.

The failing of the dgram_pair test is curious since that is an in-memory BIO and doesn't do anything with OS sockets.

**Comment by rsbeckerca** [comment]
> > There are some considerations on the platform that may apply here. Binding to specific sockets from 0-1024 are restricted to root only. I cannot change that. This applies to both UDP and TCP. I am wondering whether that is the cause of both errors above.
> 
> It seems unlikely since the same restrictions apply to Linux.
> 
> The failing of the dgram_pair test is curious since that is an in-memory BIO and doesn't do anything with OS sockets.

I can rerun the tests with V=1 if that would help. Now that I'm past 80-test_cmp_http (still failing but I have a fix for it killing perl), there might be other tests that fail but I can see the output.

**Comment by rsbeckerca** [comment]
I threw this into `gdb` at `BIO_socket`. The call is resolved to `socket(AF_INET6, SOCK_DGRAM, 17)`, which normally should work. It appears that IPv6 is not configured on the box I am testing, so we are getting `EAFNOSUPPORT`. Perhaps we need to add a check and bypass the test if IPv6 is not enabled. I am urgently working with the sysadmin to get this changed as it seems to be hitting multiple tests.

**Note:** It may take a week to get this changed because this is a shared system and I have to give lots of notice for this type of change. Would that cause problems for other pull requests that might result to get this into 3.2.0-beta?

**Comment by hlandau** [comment]
Nonetheless &mdash; I think this is a bug if our code is breaking without IPv6.

**Comment by mattcaswell** [comment]
Yeah, we absolutely must support non-IPv6 boxes. They are still common.

**Comment by mattcaswell** [comment]
Wasn't @levitte looking at similar IPv6 issues on VMS? Is this the same as those?

**Comment by t8m** [comment]
> Wasn't @levitte looking at similar IPv6 issues on VMS? Is this the same as those?

I do not think it is the same problem - in @levitte's case the IPv6 support was AFAIK completely non-present on the VMS builder, but there was still some attempt to resolve IPv6 addresses. However here the IPv6 support is enabled in headers, etc. so this builds with IPv6 support, but IPv6 is not configured, so the tests using IPv6 will fail.

So IMO completely different issue. We might want to try to configure a VM without IPv6 to test this.

**Comment by rsbeckerca** [comment]
We reconfigured to have IPv6 running on the box. This moves the problem somewhat.  As of now, `80-test_cmp_http` and `79-test_http` work. `04-test_bio_dgram` still fails, with the full log below. I am renaming this case to focus on that.
```
04-test_bio_dgram.t ..
# The results of this test will end up in test-runs/test_bio_dgram
1..1
    # Subtest: ../../test/bio_dgram_test
    1..2
        # Subtest: test_bio_dgram
        1..4
        # INFO:  @ /home/randall/openssl/test/bio_dgram_test.c:119
        # # Testing with AF_INET, local=0
        #
        ok 1 - iteration 1
        # INFO:  @ /home/randall/openssl/test/bio_dgram_test.c:125
        # # Testing with AF_INET6, local=0
        #
        ok 2 - iteration 2
        # INFO:  @ /home/randall/openssl/test/bio_dgram_test.c:119
        # # Testing with AF_INET, local=1
        #
        ok 3 - iteration 3
        # INFO:  @ /home/randall/openssl/test/bio_dgram_test.c:125
        # # Testing with AF_INET6, local=1
        #
        ok 4 - iteration 4
    ok 1 - test_bio_dgram
        # Subtest: test_bio_dgram_pair
        1..3
        # ERROR: (int) 'i >= 9' failed @ /home/randall/openssl/test/bio_dgram_test.c:565
        # [6] compared to [9]
        # 00000000:error:10000070:BIO routines:dgram_pair_write:non-fatal or transient error:/home/randall/openssl/crypto/bio/bss_dgram_pair.c:1271:
        # OPENSSL_TEST_RAND_ORDER=1694181873
        not ok 5 - iteration 1
        # ERROR: (int) 'i >= 9' failed @ /home/randall/openssl/test/bio_dgram_test.c:565
        # [6] compared to [9]
        # 00000000:error:10000070:BIO routines:dgram_pair_write:non-fatal or transient error:/home/randall/openssl/crypto/bio/bss_dgram_pair.c:1271:
        # OPENSSL_TEST_RAND_ORDER=1694181873
        not ok 6 - iteration 2
        ok 7 - iteration 3
    # OPENSSL_TEST_RAND_ORDER=1694181873
    not ok 2 - test_bio_dgram_pair
../../util/wrap.pl ../../test/bio_dgram_test => 255
not ok 1 - running bio_dgram_test

#   Failed test 'running bio_dgram_test'
#   at /home/randall/openssl/util/perl/OpenSSL/Test/Simple.pm line 77.
# Looks like you failed 1 test of 1.
Dubious, test returned 1 (wstat 256, 0x100)
Failed 1/1 subtests
```

**Comment by rsbeckerca** [comment]
The test seems to be failing at
https://github.com/openssl/openssl/blob/ebce766bb82b472eb3b796c5ee5ee2c19beb71f5/crypto/bio/bss_dgram_pair.c#L1246
with 
```
 b->rbuf
$12 = {
  start = 0x10001b310 "",
  len = 17408,
  count = 17126,
  idx = {17126, 0}
}
```
Inside, at `ring_buf_head_tail`, we hit a line on the 6th iteration, where `max_len` starts at `17408`, then hits `max_len = r->len - r->count;`, where `r->len=17408` and `r->count=17408` causing `max_len` to be `0`. This causes the test to report the above transient error. This does not seem correct.

**Comment by rsbeckerca** [comment]
I should point out that I think the DGRAM size limit is configured to be 8K on the IP stack I am using.

**Comment by t8m** [comment]
> I should point out that I think the DGRAM size limit is configured to be 65K on the IP stack I am using.

That is not the default MTU used by the dgram_pair BIO here. - it should be set to 1472. What do you see if you display the contents of the bio_dgram_pair_st inside the BIO?

**Comment by mattcaswell** [comment]
To help debug this further could you print out the values of `max_len` and `r->count` each time we enter `ring_bug_head_tail` for the first test that fails?

Also what is the value of `mtu1` in `test_bio_dgram_pair` (test/bio_dgram_test.c)

**Comment by rsbeckerca** [comment]
In `test_bio_dgram_pair`, `mtu1=1472`
Iteration 0: `max_len = 17408, r={ start = 0x10001b310 "", len = 17408, count = 0, idx = {0, 0}}`
Iteration 1: `max_len = 15344, r={ start = 0x10001b310 "", len = 17408, count = 2064, idx = {2064, 0}}`
Iteration 2: `max_len = 14901, r={ start = 0x10001b310 "", len = 17408, count = 2507, idx = {2507, 0}}`
Iteration 3: `max_len = 12837, r={ start = 0x10001b310 "", len = 17408, count = 4571, idx = {4571, 0}}`
Iteration 4: `max_len = 11777, r={ start = 0x10001b310 "", len = 17408, count = 5631, idx = {5631, 0}}`
Iteration 5: `max_len = 9713, r={ start = 0x10001b310 "", len = 17408, count = 7695, idx = {7695, 0}}`
Iteration 6: `max_len = 8620, r={ start = 0x10001b310 "", len = 17408, count = 8788, idx = {8788, 0}}`
Iteration 7: `max_len = 6556, r={ start = 0x10001b310 "", len = 17408, count = 10852, idx = {10852, 0}}`
Iteration 8: `max_len = 6392, r={ start = 0x10001b310 "", len = 17408, count = 11016, idx = {11016, 0}}`
Iteration 9: `max_len = 4328, r={ start = 0x10001b310 "", len = 17408, count = 13080, idx = {13080, 0}}`
Iteration 10: `max_len = 3176, r={ start = 0x10001b310 "", len = 17408, count = 14232, idx = {14232, 0}}`
Iteration 11: `max_len = 1112, r={ start = 0x10001b310 "", len = 17408, count = 16296, idx = {16296, 0}}`
Iteration 12: `max_len = 237, r={ start = 0x10001b310 "", len = 17408, count = 17171, idx = {17171, 0}}`
Iteration 13: `max_len = 17408, r={ start = 0x10001b310 "", len = 17408, count = 17408, idx = {17408, 0}}`
`max_len=0` at the end of this call.
In `dgram_pair_write_inner`, `total_written=237`, which seems wrong because `sizeof(hdr)=2064` and that drops through to the failure code. `saved_count=17171`. Is it possible that handling `count` is incorrect?

**Comment by rsbeckerca** [comment]
The following lines are not intuitive to me (probably my lack of understanding of the intent of the code):
https://github.com/openssl/openssl/blob/d2873946dfaff5537ea3d1adf3890e33a3f276ff/crypto/bio/bss_dgram_pair.c#L1181-L1182
Is this saying that if `sz > 0` then `dst_len = sz` but what happened to the rest of the contents of the buffer? Or is `sz` the remaining size? It might help to have that variable descriptive of what it is.

The lines following:
https://github.com/openssl/openssl/blob/d2873946dfaff5537ea3d1adf3890e33a3f276ff/crypto/bio/bss_dgram_pair.c#L1187-L1189
Seem to make the above assignment strange because `sz` will always be set to `0` if `dst_len > sz` for the next iteration of the loop.

**Comment by rsbeckerca** [comment]
> > I should point out that I think the DGRAM size limit is configured to be 65K on the IP stack I am using.
> 
> That is not the default MTU used by the dgram_pair BIO here. - it should be set to 1472. What do you see if you display the contents of the bio_dgram_pair_st inside the BIO?

@t8m I verified that the MTU is 1472. This situation is perplexing. If there is any help you can provide so I can investigate more/fix this, I would appreciate it. (see previous comments).

**Comment by mattcaswell** [comment]
> sizeof(hdr)=2064

!!!! Oh wow! That's huge! This is the cause of the problem. On my system this value is 232.

This is the line in the test that is failing:

https://github.com/openssl/openssl/blob/84f371a130dbe7a46595fbabd274f152a0e6385f/test/bio_dgram_test.c#L561-L566

The default write buffer size is 17408. Each datagram consists of a random length up to the max mtu of 1472 plus the datagram header (2064 bytes in your case).

On my system, in the worst case scenario where every datagram has the maximum length, this means we could fit `17408 / (1472 + 232) == 10.2` datagrams into the default buffer size. This test conservatively assumes that we should be able to fit at least 9 in.

But on your system the test is failing with:

````
        # ERROR: (int) 'i >= 9' failed @ /home/randall/openssl/test/bio_dgram_test.c:565
        # [6] compared to [9]
````

i.e. it only managed to fit 6 datagrams in. This seems about right. If the average datagram length is half the mtu (736). Then each datagram plus the header is `736+2064 == 2800`. With the buffer size of 17408 we can fit `17408 / 2064 == 6.2`.

**Comment by mattcaswell** [comment]
The hdr struct is defined here:

https://github.com/openssl/openssl/blob/e22ebb893e2f44bd08f69f9ce4ccfc5e4d2990e2/crypto/bio/bss_dgram_pair.c#L243-L246

I assume it is `sizeof(BIO_ADDR)` which is larger than expected (perhaps 1000 bytes assuming `sizeof(size_t)` is 64??)

a `BIO_ADDR` is simply a union containing these things:

https://github.com/openssl/openssl/blob/e22ebb893e2f44bd08f69f9ce4ccfc5e4d2990e2/include/internal/bio_addr.h#L16-L27

Where all of those types are defined by the system. It's very strange that they are so large on Non-stop!

**Comment by mattcaswell** [comment]
Possible fix for this in #22058.

@rsbeckerca - can you test the fix?

**Comment by hlandau** [comment]
My guess is what is going on here is that `struct sockaddr_un` is massive due to how NonStop decided to implement it.

TLDR: `struct sockaddr` (well, `struct sockaddr_storage`) is a tagged union structure. This means it needs to have a fixed maximum size. This is a problem with UNIX domain sockets because they are identified by paths. As I understand it Linux basically cheesed a solution to this problem by defining a fixed maximum length for UNIX domain socket paths:

```c
struct sockaddr_un {
	sa_family_t sun_family;
	char sun_path[108];
};
```

So while Linux supports long path names in most cases, for UNIX domains sockets specifically you have a special limitation in order to keep the size of `struct sockaddr_storage` reasonable.

My guess is NonStop decided they didn't care about memory usage and to choose some massive size for `sockaddr_un`, say 2k.

The obvious solution here is to base our testing on the platform's sizeof(BIO_ADDR).

Another option is to avoid using BIO_ADDR here as we never actually use UNIX domain sockets. We could just define a union of `struct sockaddr_in` and `struct sockaddr_in6`. Probably a better option but more work and more duplication. Probably not worth us doing this for Nonstop but a PR would probably be accepted if the memory wastage is considered an issue.

**Comment by hlandau** [comment]
(You could also work around this temporarily by building with -DOPENSSL_NO_UNIX_SOCK, of course)

**Comment by rsbeckerca** [comment]
> Possible fix for this in #22058.
> 
> @rsbeckerca - can you test the fix?

@mattcaswell Running a build/test cycle now. Will run the individual test then a full regression, just to make sure.

**Comment by mattcaswell** [comment]
I literally just updated the commit with the latest suggestion from @t8m...:-)

**Comment by rsbeckerca** [comment]
> I literally just updated the commit with the latest suggestion from @t8m...:-)

I just picked up this change and am restarting. No worries.

**Comment by rsbeckerca** [comment]
@mattcaswell I'm sorry to say that we only got on more iteration through the loop:
```
04-test_bio_dgram.t ..
# The results of this test will end up in test-runs/test_bio_dgram
1..1
    # Subtest: ../../test/bio_dgram_test
    1..2
        # Subtest: test_bio_dgram
        1..4
        # INFO:  @ /home/randall/openssl/test/bio_dgram_test.c:119
        # # Testing with AF_INET, local=0
        #
        ok 1 - iteration 1
        # INFO:  @ /home/randall/openssl/test/bio_dgram_test.c:125
        # # Testing with AF_INET6, local=0
        #
        ok 2 - iteration 2
        # INFO:  @ /home/randall/openssl/test/bio_dgram_test.c:119
        # # Testing with AF_INET, local=1
        #
        ok 3 - iteration 3
        # INFO:  @ /home/randall/openssl/test/bio_dgram_test.c:125
        # # Testing with AF_INET6, local=1
        #
        ok 4 - iteration 4
    ok 1 - test_bio_dgram
        # Subtest: test_bio_dgram_pair
        1..3
        ok 5 - iteration 1
        # ERROR: (int) 'i >= 9' failed @ /home/randall/openssl/test/bio_dgram_test.c:568
        # [7] compared to [9]
        # 00000000:error:10000070:BIO routines:dgram_pair_write:non-fatal or transient error:/home/randall/openssl/crypto/bio/bss_dgram_pair.c:1272:
        # OPENSSL_TEST_RAND_ORDER=1694446579
        not ok 6 - iteration 2
        ok 7 - iteration 3
    # OPENSSL_TEST_RAND_ORDER=1694446579
    not ok 2 - test_bio_dgram_pair
```

**Comment by mattcaswell** [comment]
Well, at least we got one more iteration further :-)

Will take another look tomorrow.

**Comment by rsbeckerca** [comment]
> Well, at least we got one more iteration further :-)
> 
> Will take another look tomorrow.

Sounds good to me. I'm probably not gong to be able to do much about it tomorrow - at the NonStop TBC Conference talking about Ransomware Attack Vectors tomorrow.

**Comment by rsbeckerca** [comment]
@mattcaswell I can't decipher what, but my gut is telling me that there is a missing loop termination condition causing the lengths to reset, which then causes the error. I also cannot figure out why the header appears to large on NonStop. There is one pad in the old sockaddr (not used) that is 64 bytes. This is perplexing to me also. It should not be 2062.

**Comment by rsbeckerca** [comment]
> NO_UNIX_SOCK

I think this might break access to PRNGD, which is accessed via UNIX sockets. Going through TCP is an option, but a painful one. We only use PRNGD on ia64, not x86, though.

**Comment by mattcaswell** [comment]
I have added another commit to #22058 to address the failure in the other iteration. I was actually able to recreate this on my local machine by artificially increasing the size of the `BIO_ADDR` structure.

**Comment by mattcaswell** [comment]
@rsbeckerca - please could you retest?

**Comment by rsbeckerca** [comment]
@mattcaswell Hurrah!
```
04-test_bio_dgram.t ..
# The results of this test will end up in test-runs/test_bio_dgram
1..1
    # Subtest: ../../test/bio_dgram_test
    1..2
        # Subtest: test_bio_dgram
        1..4
        # INFO:  @ /home/randall/openssl/test/bio_dgram_test.c:120
        # # Testing with AF_INET, local=0
        #
        ok 1 - iteration 1
        # INFO:  @ /home/randall/openssl/test/bio_dgram_test.c:126
        # # Testing with AF_INET6, local=0
        #
        ok 2 - iteration 2
        # INFO:  @ /home/randall/openssl/test/bio_dgram_test.c:120
        # # Testing with AF_INET, local=1
        #
        ok 3 - iteration 3
        # INFO:  @ /home/randall/openssl/test/bio_dgram_test.c:126
        # # Testing with AF_INET6, local=1
        #
        ok 4 - iteration 4
    ok 1 - test_bio_dgram
        # Subtest: test_bio_dgram_pair
        1..3
        ok 5 - iteration 1
        ok 6 - iteration 2
        ok 7 - iteration 3
    ok 2 - test_bio_dgram_pair
../../util/wrap.pl ../../test/bio_dgram_test => 0
ok 1 - running bio_dgram_test
ok
All tests successful.
Files=1, Tests=1,  1 wallclock secs ( 0.01 usr  0.00 sys +  0.63 cusr  0.00 csys =  0.64 CPU)
Result: PASS
```

**Comment by mattcaswell** [comment]
Great! Are there any other outstanding NonStop test failures - or do you get a clean run now?

**Comment by rsbeckerca** [comment]
> Great! Are there any other outstanding NonStop test failures - or do you get a clean run now?

Restarting the CI stream. Need a couple of hours but it should be clean.

**Comment by rsbeckerca** [comment]
> > Great! Are there any other outstanding NonStop test failures - or do you get a clean run now?
> 
> Restarting the CI stream. Need a couple of hours but it should be clean.

@mattcaswell I am pulling master, merging your fix, and running the NonStop CI.

**Comment by rsbeckerca** [comment]
@mattcaswell Clean run!

**Comment by mattcaswell** [comment]
Great! Thanks for confirming.

**Comment by veeranji11** [comment]
@rsbeckerca / @mattcaswell 
I am.also.facing same issue , can you please provide the fix to resolve this issue.

Thanks ,
Veeranji

**Comment by veeranji11** [comment]
I am trying to install the same version and  facing the same error . If you could remember the fix ,kindly share with me.

If anyone is aware of the fix . Kindly share

**Comment by rsbeckerca** [comment]
> @rsbeckerca I am trying to install the same version and facing the same error . If you could remember the fix ,kindly share with me.
> 
> If anyone is aware of the fix . Kindly share

Is there a reason you need the `openssl-3.2-alpha1` version instead of `openssl-3.2` ?

**Comment by veeranji11** [comment]
@rsbeckerca 
I am not installing alpha1, where as same issue appeared even for 3.2 sha .

**Comment by rsbeckerca** [comment]
> @rsbeckerca I am not installing alpha1, where as same issue appeared even for 3.2 sha .

Can you open a new issue? This is working correctly on NonStop. We would need to see your configuration and the test output.

**Comment by rsbeckerca** [comment]
@veeranji11 to clarify, this particular issue is closed as resolved. In order to work on your situation, we need a new case opened, with details about how you built the code and the errors you are seeing - they may be the same or different, but the configuration is crucial. Use the following command at the highest directory in your clone (probably `openssl`):
```
perl configdata.pm --dump
```
That will tell the team how you built the code. Also, please send the output of the following command:
```
git show --oneline -s
```
which will show the commit you are using for the build. Also include the errors.

Please do this using the `New Issue` button from the `Issues` tab in the `openssl/openssl` repository.
